GVFS/GVFS.Common/NuGetUpgrader/NuGetUpgrader.cs

```diff
@@ -0,0 +1,466 @@
+            ITracer tracer,
+            bool dryRun,
+            bool noVerify,
+            out string error)
+        {
+            NugetUpgraderConfig upgraderConfig = new NugetUpgraderConfig(tracer, new LocalGVFSConfig());
+            bool isConfigured;
+            bool isEnabled;
+
+            if (!upgraderConfig.TryLoad(out isEnabled, out isConfigured, out error))
+            {
+                if (isEnabled && !isConfigured)
+                {
+                    tracer.RelatedWarning($"NuGetUpgrader is enabled, but is not properly configured. Error: {error}");
+
+                    return new NuGetUpgrader(
+                        ProcessHelper.GetCurrentProcessVersion(),
+                        tracer,
+                        upgraderConfig,
+                        dryRun,
+                        ProductUpgraderInfo.GetAssetDownloadsPath(),
+                        personalAccessToken: null);
+                }
+
+                return null;
+            }
+
+            if (!TryGetPersonalAccessToken(
+                GitBinPath,
+                upgraderConfig.FeedUrlForCredentials,
+                tracer,
+                out string token,
+                out error))
+            {
+                tracer.RelatedWarning($"NuGetUpgrader was not able to acquire Personal Access Token to access NuGet feed. Error: {error}");
+            }
+
+            NuGetUpgrader upgrader = new NuGetUpgrader(
+                ProcessHelper.GetCurrentProcessVersion(),
+                tracer,
+                upgraderConfig,
+                dryRun,
+                ProductUpgraderInfo.GetAssetDownloadsPath(),
+                token);
+
+            return upgrader;
+        }
+
+        public void Dispose()
+        {
+            this.nuGetFeed?.Dispose();
+        }
+
+        public bool UpgradeAllowed(out string message)
+        {
+            if (string.IsNullOrEmpty(this.nugetUpgraderConfig.FeedUrl))
+            {
+                message = "Nuget Feed URL has not been configured";
+                return false;
+            }
+            else if (string.IsNullOrEmpty(this.nugetUpgraderConfig.PackageFeedName))
+            {
+                message = "URL to lookup credentials has not been configured";
+                return false;
+            }
+            else if (string.IsNullOrEmpty(this.nugetUpgraderConfig.FeedUrlForCredentials))
+            {
+                message = "URL to lookup credentials has not been configured";
+                return false;
+            }
+            else
+            {
+                message = null;
+            }
+
+            message = null;
+            return true;
+        }
+
+        public bool TryQueryNewestVersion(out Version newVersion, out string message)
+        {
+            try
+            {
+                IList<IPackageSearchMetadata> queryResults = this.nuGetFeed.QueryFeedAsync(this.nugetUpgraderConfig.PackageFeedName).GetAwaiter().GetResult();
```

Comment:
Only the NuGetUpgrader version is asynchronous, the original method GitHubUpgrader is not. I would prefer to keep to the original contract, rather than make the original version match an async API. As none of our APIs are Async, I didn't see an advantage to making this expose an async implementation - thoughts?

Reply:
Sorry, what I meant was why does the `NuGetFeed` expose an async method instead of it having a synchronous call and hide that the NuGet API is async instead of here.

Reply:
ah - I see. That is another option. I chose the NuGetUpgrader as the layer to absorb that detail, as it is the orchestrating these calls and is the main API that consumers will consume at this point.